### ISSUE TITLE
more warnings about unnecessary casts

### ISSUE THREAD

**Issue by oconnor663** [rust-lang-nursery/rust-clippy#2100]
Currently if I write this egregious line, Clippy warns me about the unnecessary cast:

```rust
let a = 5usize as usize;
```

```
warning: casting to the same type is unnecessary (`usize` -> `usize`)          
 --> src/main.rs:2:13                  
  |                                    
2 |     let a = 5usize as usize;       
  |             ^^^^^^^^^^^^^^^        
  |                                    
  = note: #[warn(unnecessary_cast)] on by default                              
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.165/index.html#unnecessary_cast                                   
```

That's great, but there are some very similar cases that it misses:

```rust
let a: usize = 5 as usize;
let b: usize = a as usize;
```

Probably both of those should generate a warning, the first because the literal will assume the right type automatically, and the second because `a` is already `usize`.

**Comment by oli-obk** [comment]
Tracking explicit type annotations is rather hard. This will need a full function analysis. Anyone taking this issue should remember to add tests for `allow` attributes at different places inside the function.

**Comment by matthiaskrgr** [comment]
Found a similar case that also doesn't trigger any warning for me (clippy 0.0.176)
````rust
let useless_cast: u64 = (4 as u64) as u64;
````

**Comment by matthiaskrgr** [comment]
For the record, ````#![warn(trivial_numeric_casts)]```` which comes from rustc reports all of these, so I guess we can close this.

**Comment by matthiaskrgr** [comment]
````rust
#![warn(trivial_numeric_casts)]

fn main() {
    let a: usize = 5 as usize;
    let b: usize = a as usize;

    let _useless_cast: u64 = (4 as u64) as u64;
}
````

````
warning: trivial numeric cast: `usize` as `usize`
 --> src/main.rs:4:20
  |
4 |     let a: usize = 5 as usize;
  |                    ^^^^^^^^^^
  |
note: lint level defined here
 --> src/main.rs:1:9
  |
1 | #![warn(trivial_numeric_casts)]
  |         ^^^^^^^^^^^^^^^^^^^^^
  = help: cast can be replaced by coercion; this might require a temporary variable
warning: trivial numeric cast: `usize` as `usize`
 --> src/main.rs:5:20
  |
5 |     let b: usize = a as usize;
  |                    ^^^^^^^^^^
  |
  = help: cast can be replaced by coercion; this might require a temporary variable
warning: trivial numeric cast: `u64` as `u64`
 --> src/main.rs:7:30
  |
7 |     let _useless_cast: u64 = (4 as u64) as u64;
  |                              ^^^^^^^^^^
  |
  = help: cast can be replaced by coercion; this might require a temporary variable
warning: trivial numeric cast: `u64` as `u64`
 --> src/main.rs:7:30
  |
7 |     let _useless_cast: u64 = (4 as u64) as u64;
  |                              ^^^^^^^^^^^^^^^^^
  |
  = help: cast can be replaced by coercion; this might require a temporary variable
warning: unused variable: `b`
 --> src/main.rs:5:9
  |
5 |     let b: usize = a as usize;
  |         ^ help: consider using `_b` instead
  |
  = note: #[warn(unused_variables)] on by default
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
````
